asset/asset.go

```diff
@@ -943,6 +944,447 @@ type GroupKeyReveal interface {
+//     flexibility for custom tapscript subtrees.
+//
+// User-defined script spending leaves are nested under
+// `[tweaked_custom_branch]` as a single node hash, `custom_root_hash`. This
+// hash may represent either a single leaf or the root hash of an entire
+// subtree.
+//
+// If `custom_root_hash` is not provided, it defaults to a 32-byte zero-filled
+// array. In this case, no valid script spending path can correspond to the
+// custom subtree root hash due to the pre-image resistance of SHA-256.
+//
+// A sibling node is included alongside the `custom_root_hash` node. This
+// sibling is a non-spendable script leaf containing `[OP_RETURN]`. Its
+// presence ensures that one of the two positions in the first layer of the
+// tapscript tree is occupied by a branch node. Due to the pre-image
+// resistance of SHA-256, this prevents the existence of a second recognizable
+// genesis asset ID leaf.
+//
+// The final tapscript tree adopts the following structure:
+//
+//	                       [tapscript_root]
+//	                         /          \
+//	[OP_RETURN <genesis asset ID>]   [tweaked_custom_branch]
+//	                                      /        \
+//	                              [OP_RETURN]   <custom_root_hash>
+//
+// Where:
+//   - [tapscript_root] is the root of the final tapscript tree.
+//   - [OP_RETURN <genesis asset ID>] is a first-layer non-spendable script
+//     leaf that commits to the genesis asset ID.
+//   - [tweaked_custom_branch] is a branch node that serves two purposes:
+//     1. It cannot be misinterpreted as a genesis asset ID leaf.
+//     2. It optionally includes user-defined script spending leaves.
+//   - <custom_root_hash> is the root hash of the custom tapscript subtree.
+//     If not specified, it defaults to a 32-byte zero-filled array.
+//   - [OP_RETURN] is a non-spendable script leaf containing the script
+//     `OP_RETURN`. Its presence ensures that [tweaked_custom_branch] remains
+//     a branch node and cannot be a valid genesis asset ID leaf.
+type GroupKeyRevealTapscript struct {
+	// root is the final tapscript root after all tapscript tweaks have
+	// been applied. The asset group key is derived from this root and the
+	// internal key.
+	root chainhash.Hash
+
+	// customSubtreeRoot is an optional root hash representing a
+	// user-defined tapscript subtree that is integrated into the final
+	// tapscript tree. This subtree may define script spending conditions
+	// associated with the group key.
+	customSubtreeRoot fn.Option[chainhash.Hash]
+
+	// customSubtreeInclusionProof provides the inclusion proof for the
+	// custom tapscript subtree. It is required to spend the custom
+	// tapscript leaves within the tree.
+	//
+	// NOTE: This field should not be serialized as part of the group key
+	// reveal. It is included here to ensure it is constructed concurrently
+	// with the tapscript root, maintaining consistency and minimizing
+	// errors.
+	customSubtreeInclusionProof []byte
```

Comment:
Any reason this isn't just `txscript.ControlBlock`? 

Why wouldn't it be serialized as part of the reveal? It includes the internal key and also the inclusion proof all the way up.

Reply:
My thinking here is not to include `txscript.ControlBlock` in the serialization to avoid redundancy. Breakdown:

- **`InternalKey`**: Already serialized separately.  
- **`OutputKeyYIsOdd`**: Derivable from internal key and tapscript root.  
- **`LeafVersion`**: Typically constant; serialization might not be necessary?
- **`InclusionProof`**: Derivable from the genesis asset ID which should be available separately whenever group key reveal is at hand.  

I'm aiming to eliminate data duplication, ensuring that each field serves as a definitive "source of truth." And also reduce serialization size.

I suppose if the default value of `LeafVersion` changes we could version bump group key reveal? Maybe we should just include it now?